enhancements/storage/cleanup-lso-symlink.md

```diff
@@ -0,0 +1,66 @@
+---
+title: handle-local-volume-deletion
+authors:
+  - "@gnufied"
+reviewers:
+  - "@jsafrane”
+  - "@fbertina"
+  - "@chuffman"
+approvers:
+  - "@..."
+creation-date: 2021-04-17
+last-updated: 2021-04-17
+status: implementable
+see-also: https://github.com/openshift/enhancements/blob/master/enhancements/storage/cleanup-lso-symlink.md
+replaces:
+superseded-by:
+---
+
+# Cleanup and handling of LocalVolume and LocalVolumeSet deletion
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes a mechanism that will allow PVs provisioned by local-storage-operator(LSO) to be deleted and corresponding symlinks to be removed from the node when `LocalVolume` or `LocalVolumeSet` objects are deleted.
+
+## Motivation
+
+Currently when a user deletes `LocalVolume` or `LocalVolumeSet` objects - we do not automatically delete the PVs created from those objects and neither do we remove the symlinks created on the node. This has been documented as a limitation because removal of symlinks and PVs can be sometimes race-prone.
+
+This design fixes that problem and ensures that both PVs and symlinks are removed when `LocalVolume` and `LocalVolumeSet` objects are removed.
+
+### Goals
+
+* Do not allow deletion of `LocalVolumeSet` object if it has bound PVCs.
+* Remove unbound PVs and remove symlinks on the host if `LocalVolume` or `LocalVolumeSet` objects are removed.
+
+### Non-Goals
+
+* There is another use-case that this design does not solve is - editing of `LocalVolume` and `LocalVolumeSet` objects. If user edits LSO CRs in such a manner that - a node which was previously selected is no longer included via
+node-selector or if CR excludes a device which was previously used for local-storage then - created PV is not automatically removed and corresponding symlink on the node is not removed either.
+
+## Proposal
+
+1. We propose that when `LocalVolume` or `LocalVolumeSet` object is created - they both contain a finalizer (`LocalVolume` object already has a finalizer). In addition we propose that - when user deletes these objects then finalizer can only be removed:
+   a. If there are no bound PVs that are created by this LSO CR.
+   b. None of the existing PVs(that match LSO CR) should have reclaim policy of `Retain.`
+   c. If there are unbound PVs with reclaimPolicy of `Delete` then finalizer should not be removed until all such PVs are deleted.
```

Comment:
I admit that the conditions here (and below) are hard to read and put together.
In a. it allows deletion when there are unbound PVs and in b. and c. it does not allow to delete those.

In addition, why should ReclaimPolicy matter? Unound PV was not used, so there is no data to retain on them.

Perhaps it would be better to use PV phase:

* LSO removes the finalizer on LV/LVS only when all PVs associated with it are deleted.
* LSO automatically deletes Pending and Available PVS for LV/LVS with a deletion timestamp, as we expect there is no data on them. See below how LSO deletes them avoiding potential races with PV controller.
* Released PVs with `Delete` policy are deleted according to the policy.
* Bound PVs or Released PVs with `Retain` policy actively block LSO from removing the finalizer. An event is sent to tell users which PV blocks the deletion.
  * For Bound PVs, user should delete the corresponding PVC to get the PV Released.
  * For Released PVs with reclaim policy "Retain", user should back up any important data from the volume and delete the PV.

Reply:
I reworded this section.